src/compiler/checker.ts

```diff
@@ -8987,6 +8978,45 @@ module ts {
             }
         }
 
+        function checkElementTypeOfArrayOrString(arrayOrStringType: Type, expressionForError: Expression): Type {
+            Debug.assert(languageVersion < ScriptTarget.ES6);
+            var isJustString = allConstituentTypesHaveKind(arrayOrStringType, TypeFlags.StringLike);
+            
+            // Check isJustString because removeTypesFromUnionType will only remove types if it doesn't result

```

Comment:
I feel like this behavior should be in removeTypesFromUnionType.  i.e. have a flag like 'allowAllTypesToBeRemoved'.   It's very weird that 'removeTypesFromUnionType' will choose to not actually remove the type, and that you have to work around that here.